File: hospital-ai/backend/app/schemas/intake.py
"""Intake Agent API schemas."""

from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, Field


class IntakeProcessRequest(BaseModel):
    document_id: UUID
    # Optional explicit linkage checks
    patient_id: Optional[UUID] = None
    appointment_id: Optional[UUID] = None
    doctor_id: Optional[UUID] = None


class IntakeProcessResponse(BaseModel):
    job_id: str
    patient_id: str
    document_id: str
    status: str
    patient_context: Dict[str, Any]
    entities: Dict[str, Any]
    risk_profile: Dict[str, Any]
    knowledge_graph: Dict[str, Any]
    ocr: Dict[str, Any]
    timeline: List[Dict[str, Any]]
    confidence_score: float
    diagnosis_request_preview: Optional[Dict[str, Any]] = None


class DocumentJobResponse(BaseModel):
    id: UUID
    patient_id: UUID
    medical_record_id: Optional[UUID] = None
    document_id: Optional[UUID] = None
    appointment_id: Optional[UUID] = None
    doctor_id: Optional[UUID] = None
    status: str
    current_step: Optional[str] = None
    progress_pct: int = 0
    ocr_provider: Optional[str] = None
    ocr_text: Optional[str] = None
    ocr_confidence: Optional[float] = None
    extracted_entities: Optional[Any] = None
    risk_profile: Optional[Any] = None
    knowledge_graph_refs: Optional[Any] = None
    patient_context: Optional[Any] = None
    timeline: Optional[Any] = None
    error_message: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    completed_at: Optional[datetime] = None


class DocumentJobListResponse(BaseModel):
    items: List[DocumentJobResponse]
    total: int


class PatientAIContextResponse(BaseModel):
    id: UUID
    patient_id: UUID
    context_version: int
    context_json: Dict[str, Any]
    medical_history_summary: Optional[str] = None
    risk_level: Optional[str] = None
    risk_profile: Optional[Any] = None
    confidence_score: Optional[float] = None
    knowledge_graph_id: Optional[UUID] = None
    source_job_ids: Optional[List[UUID]] = None
    source_document_ids: Optional[List[UUID]] = None
    built_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class KnowledgeGraphResponse(BaseModel):
    id: UUID
    patient_id: UUID
    graph_version: int
    nodes: List[Any] = Field(default_factory=list)
    edges: List[Any] = Field(default_factory=list)
    summary: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class IntakeDashboardResponse(BaseModel):
    patient_id: str
    context: Optional[PatientAIContextResponse] = None
    knowledge_graph: Optional[KnowledgeGraphResponse] = None
    jobs: List[DocumentJobResponse] = Field(default_factory=list)
    documents: List[Dict[str, Any]] = Field(default_factory=list)

File: hospital-ai/frontend/src/types/intake.ts
/** Intake Agent types — Patient Context is the sole downstream input. */

export type DocumentJobStatus =
  | 'Queued'
  | 'Validating'
  | 'OCR'
  | 'EntityExtraction'
  | 'RiskProfiling'
  | 'KnowledgeGraph'
  | 'ContextBuild'
  | 'Completed'
  | 'Failed';

export interface TimelineStep {
  step: string;
  detail: string;
  at: string;
}

export interface DocumentProcessingJob {
  id: string;
  patient_id: string;
  medical_record_id?: string | null;
  document_id?: string | null;
  appointment_id?: string | null;
  doctor_id?: string | null;
  status: DocumentJobStatus | string;
  current_step?: string | null;
  progress_pct: number;
  ocr_provider?: string | null;
  ocr_text?: string | null;
  ocr_confidence?: number | null;
  extracted_entities?: Record<string, unknown> | null;
  risk_profile?: Record<string, unknown> | null;
  knowledge_graph_refs?: Record<string, unknown> | null;
  patient_context?: Record<string, unknown> | null;
  timeline?: TimelineStep[] | null;
  error_message?: string | null;
  created_at: string;
  updated_at: string;
  completed_at?: string | null;
}

export interface PatientAIContext {
  id: string;
  patient_id: string;
  context_version: number;
  context_json: Record<string, unknown>;
  medical_history_summary?: string | null;
  risk_level?: string | null;
  risk_profile?: Record<string, unknown> | null;
  confidence_score?: number | null;
  knowledge_graph_id?: string | null;
  source_job_ids?: string[] | null;
  source_document_ids?: string[] | null;
  built_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface KnowledgeGraphPreview {
  id: string;
  patient_id: string;
  graph_version: number;
  nodes: Array<Record<string, unknown>>;
  edges: Array<Record<string, unknown>>;
  summary?: string | null;
  created_at: string;
  updated_at: string;
}

export interface IntakeDocumentMeta {
  id: string;
  medical_record_id: string;
  file_name: string;
  file_type?: string;
  file_size?: number;
  created_at: string;
}

export interface IntakeDashboard {
  patient_id: string;
  context: PatientAIContext | null;
  knowledge_graph: KnowledgeGraphPreview | null;
  jobs: DocumentProcessingJob[];
  documents: IntakeDocumentMeta[];
}

export interface IntakeProcessResult {
  job_id: string;
  patient_id: string;
  document_id: string;
  status: string;
  patient_context: Record<string, unknown>;
  entities: Record<string, unknown>;
  risk_profile: Record<string, unknown>;
  knowledge_graph: Record<string, unknown>;
  ocr: Record<string, unknown>;
  timeline: TimelineStep[];
  confidence_score: number;
  diagnosis_request_preview?: Record<string, unknown> | null;
}
